draft-ietf-mls-protocol.md

```diff
@@ -1435,48 +1461,213 @@ all arrive at the following state:
+
+### Updating the Application Secret
+
+The following rules apply to an Application Secret:
+
+- Senders MUST only use the Application Secret once and monotonically
+  increment the generation of their secret.
+
+- Receivers MUST delete an Application Secret once it has been used to
+  derive the corresponding AEAD key and nonce as well as the next Application
+  Secret. Receivers MAY keep the AEAD key and nonce around for some
+  reasonable period.
 
-For every epoch, the root key of the ratcheting tree can be used to
-derive key material for symmetric operations such as encryption/AEAD and MAC;
-AEAD or MAC MUST be used to ensure that the message originated from a member
-of the group.
+- Receivers MUST delete AEAD keys and nonces once they have been used to
+  successfully decrypt a message.
 
-In addition, asymmetric signatures SHOULD be used to authenticate the sender
-of a message.
+### Application AEAD Key Calculation
 
-In combination with server-side enforced ordering, data from previous messages
-is used (as a salt when hashing) to:
+The Application AEAD keying material is generated from the following
+input values:
 
- * add freshness to derived symmetric keys
- * cryptographically bind the transcript of all previous messages with the current
-   group shared secret
+- The Application Secret value;
+- A purpose value indicating the specific value being generated;
+- The length of the key being generated.
 
-Possible candidates for that are:
+Note, that because the identity of the participant using the keys to send data
+is included in the initial Application Secret, all successive updates to the
+Application secret will implicitly inherit this ownership.
 
- * the key used for the previous message (hash ratcheting)
- * the counter of the previous message (needs to be known to new members of the group)
- * the hash of the previous message (proof that other participants saw the same history)
+All the traffic keying material is recomputed whenever the underlying
+Application Secret changes.
 
-The requirement for this is that all participants know these values.
-If additional clear-text fields are attached to messages (like the counter), those
-fields MUST be protected by a signed message envelope.
 
-Alternatively, the hash of the previous message can also be included as an additional
-field rather than change the encryption key. This allows for a more flexible approach,
-because the receiving party can choose to ignore it (if the value is not known, or if
-transcript security is not required).
+## Message Encryption and Decryption
+
+The Group participants MUST use the AEAD algorithm associated with
+the negotiated MLS ciphersuite to AEAD encrypt and decrypt their
+Application messages and sign them as follows:
+
+~~~~~
+    struct {
+        opaque content<0..2^32-1>;
+        opaque signature<0..2^16-1>;
+        uint8 zeros[length_of_padding];
+    } MLSPlaintext;
+
+    struct {
+        uint8  group[32];
```

Comment:
This isn't necessary.  In practice, it will be provided by application context, and it will be authenticated by including it under the message signature.

Reply:
This is because you think of it as an ID for the conversation, while I think of it as an ID over the conversation AND a specific set of users. Also, I don't think we should depend on the Application for this but the opposite.